client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -525,6 +525,116 @@ private RetryAndThrowConditionally createRetryConfig(final ControllerImplConfig
         });
     }
 
+    @Override
+    public CompletableFuture<Boolean> addSubscriber(String scope, String streamName, String subscriber) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkNotNull(subscriber, "subscriberId");
+        final long requestId = requestIdGenerator.get();
+        long traceId = LoggerHelpers.traceEnter(log, "addSubscriber", subscriber, requestId);
+
+        final CompletableFuture<AddSubscriberStatus> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<AddSubscriberStatus> callback = new RPCAsyncCallback<>(requestId, "addSubscriber", scope, streamName, subscriber);
+            new ControllerClientTagger(client, timeoutMillis).withTag(requestId, "addSubscriber", scope, streamName)
+                    .addSubscriber(ModelHelper.decode(scope, streamName, subscriber), callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(x -> {
+            switch (x.getStatus()) {
+                case FAILURE:
+                    log.warn(requestId, "Failed to update stream: {}", streamName);
+                    throw new ControllerFailureException("Failed to add subscriber: " + subscriber);
+                case SCOPE_NOT_FOUND:
+                    log.warn(requestId, "Scope not found: {}", scope);
+                    throw new IllegalArgumentException("Scope does not exist: " + scope);
+                case STREAM_NOT_FOUND:
+                    log.warn(requestId, "Stream does not exist: {}", streamName);
+                    throw new IllegalArgumentException("Stream does not exist: " + streamName);
+                case SUBSCRIBER_EXISTS:
+                    log.warn(requestId, "Subscriber {} for stream {}/{} already exists {}.", subscriber, scope, streamName);
```

Comment:
2 unrelated users could try to add the same subscriber name/Id and if we just return success on each addition they'll never know they are overwriting each other's data.
So it is important to error out if the subscriber already exists.

Reply:
i see your point. however let me ask a counter question - is the subscriber id same as readergroup id or is it chosen by the user?
if its former, then it doesnt matter if two users were attempting to perform this action simultaneously. 
However, if its implicit and handled at a readergroup level,

Reply:
same as readergroup id - it is fine for 2 users to do this simultaneously, but only 1 would succeed and the other would fail.... a subsequent getSubscribers() can be used to determine of the subscriber actually got added.....

Reply:
i am not convinced if we should throw an exception in this case, because its not an exceptional but idempotent case. return a boolean to indicate if it exists.. throwing illegal argument is certainly misleading.

Reply:
changed return status for SUBSCRIBER_EXISTS to false instead of throwing exception.